src/css3/new_tokenizer.rs

```diff
@@ -0,0 +1,161 @@
+// note: input_stream should come from a shared lib.
+use crate::html5_parser::input_stream::InputStream;
+
+#[derive(Debug, PartialEq)]
+pub enum NumberKind {
+    Number,
+    Integer,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Number {
+    kind: NumberKind,
+    value: f32,
+}
+
+impl Number {
+    pub fn new(kind: NumberKind, value: f32) -> Number {
+        Number { kind, value }
+    }
+}
+
+/// CSS Tokenizer according to the [w3 specification](https://www.w3.org/TR/css-syntax-3/#tokenization)
+pub struct Tokenizer<'stream> {
+    pub stream: &'stream mut InputStream,
+    /// Next token that has been consumed.
+    pub current_token: Option<String>,
+    /// Next token that has not yet been consumed
+    pub next_token: Option<String>,
+}
+
+impl<'stream> Tokenizer<'stream> {
+    pub fn new(stream: &'stream mut InputStream) -> Tokenizer {
+        Tokenizer {
+            stream,
+            current_token: None,
+            next_token: None,
+        }
+    }
+
+    /// 4.3.2. [Consume comments](https://www.w3.org/TR/css-syntax-3/#consume-comment)
+    pub fn consume_comment(&mut self) {
+        if self.stream.current_char().utf8() == '/' && self.stream.next_char().utf8() == '/' {
+            while self.stream.current_char().utf8() != '*' && self.stream.next_char().utf8() != '/'
+            {
+                self.stream.read_char();
+            }
+
+            // consume '*/'
+            self.consume_chars(2);
+        };
+    }
+
+    /// 4.3.12. [Consume a number](https://www.w3.org/TR/css-syntax-3/#consume-number)
+    pub fn consume_number(&mut self) -> Number {
+        let mut value = String::new();
+        let mut kind = NumberKind::Integer;
+        let lookahead = self.stream.current_char().utf8();
+
+        if lookahead == '+' || lookahead == '-' {
+            value.push(self.stream.read_char().utf8());
+        }
+
+        value.push_str(&self.consume_digits());
+
+        if self.stream.current_char().utf8() == '.' && self.stream.next_char().utf8().is_numeric() {
+            value.push_str(&self.consume_chars(2));
+            kind = NumberKind::Number;
+        }
+
+        // type should be "number"
+        value.push_str(&self.consume_digits());
+
+        // todo: move them to gobal constants
+        // U+0045: LATIN CAPITAL LETTER E (E)
+        // U+0065: LATIN SMALL LETTER E (e)
+        if self.stream.current_char().utf8() == '\u{0045}'
+            || self.stream.current_char().utf8() == '\u{0065}'
+        {
+            value.push(self.stream.read_char().utf8());
+            kind = NumberKind::Number;
+
+            if self.stream.current_char().utf8() == '-' || self.stream.current_char().utf8() == '+'
+            {
+                value.push(self.stream.read_char().utf8());
+            }
+        }
+
+        value.push_str(&self.consume_digits());
+
+        Number::new(kind, value.parse().unwrap())
```

Comment:
Have you put thought into how errors will percolate up to the caller?

Reply:
Ok, I understand now.

In cases where we know that the call will succeed, despite the `.unwrap()`, how about using `.expect()` instead, to communicate to developers that this call will always succeed?

Reply:
Any error message that will allow the developer to find the code again, e.g., this (or anything):

```rust
let number = value.parse().expect("error parsing number");
Number::new(kind, number)
```

How about all lowercase, and no capital letters?